reps/2022-04-01-resource-control.md

```diff
@@ -0,0 +1,198 @@
+
+The reason of using systemd:
+- Systemd is highly recommended by [runc](https://github.com/opencontainers/runc/blob/main/docs/cgroup-v2.md.) for cgroup v2.
+- If we use cgroupfs in a systemd based system, there will be more than one component which simultaneously manages the cgroup tree.
+- In the systemd [docs](https://www.freedesktop.org/wiki/Software/systemd/ControlGroupInterface/), we can know that, in future, create/delete cgroups will be unavailable to userspace applications, unless done via systemd's APIs.
+- The systemd has a good abstract API of cgroups. 
+
+So, we should support both cgroupfs and systemd. And we will provide a environment variable named `RAY_CGROUP_MANAGER` which is used to change the cgroup manager in cluster level.
+
+### Changes needed in Ray
+
+1. Add a cgroup runtime environment plugin
+
+The cgroup plugin is used to create or delete cgroups, and bind worker processes to cgroups. It is loaded in the Agent. 
+
+The cgroup plugin should expose an abstract interface that can hide the differences between cgroupfs/systemd and cgroup v1/v2.
+
+2. Generate the command line of cgroup.
+
+Agent should generate the command line from the cgroup plugin. The command line will be appended to the `command_prefix` field of runtime env context. A command line sample is like:
+
+```
+mkdir /sys/fs/cgroup/{worker_id} && echo "200000 1000000" > /sys/fs/cgroup/foo/cpu.max && echo {pid} > /sys/fs/cgroup/foo/cgroup.procs
+```
+
+3. Start the worker process with the command line.
+
+Currently, we use `setup_worker.py` to enforce the runtime environments. `setup_worker.py` will merge the `command_prefix` of runtime env context and real worker command. In the same way, the command about cgroup will be run with the worker process.
+
+
+4. The enhancement of runtime environment pluggability
+
+When worker processes die, we should delete the cgroup which is created for the processes. This work is only needed for cgroupfs because systemd can automatically delete the cgroup when the process dies.
+
+So, we should make some changes to allow each runtime environment plugin to do some work when worker processes die.
+- Add `worker_id` field to the existing PRC request `GetOrCreateRuntimeEnvRequest` and `DeleteRuntimeEnvIfPossibleRequest`. These two RPCs are used to create and delete runtime environments from the Raylet to the Agent.
+- Add two methods, named `create_for_worker` and `delete_for_worker`, in the plugin interface. We already have two methods named `create` and `delete` in the interface. The difference is that, `delete` is called only when there is no worker using the runtime environment(The reference counting to be zero), but `delete_for_worker` is called every time when a process dies.
+
+Another change is that we should support loading runtime env plugins when we start the Ray nodes. For example:
+```
+ray start --head --runtime-env-plugins="ray.experimental.cgroup_env_plugin:ray.experimental.test_env_plugin"
+```
```

Comment:
Is this different from `eager_install`, which already exists in Ray?

Reply:
Two things absolutely. This option is used to register plugins into the dashboard agent. For example, you can register a plugin named "conda" when start the ray node. So, when a job uses "conda" field  in `runtime_env` after the ray cluster starts, agent could call the "conda"  plugin directly to set up the environment.

Reply:
Makes sense!  In the [pluggability design](https://docs.google.com/document/d/1x1JAHg7c0ewcOYwhhclbuW0B0UC7l92WFkF4Su0T-dk/edit?usp=sharing) there's a related idea which is to use environment variables, e.g. `RAY_RUNTIME_ENV_PLUGIN_CGROUP=ray.experimental.test_cgroup_plugin ray start --head`.  These both seem straightforward to implement but I'm not sure which approach is the best, do you have any thoughts about this?